self_unenrollment_tool/views.py

```diff
@@ -0,0 +1,252 @@
+
+
+@require_lti_launch
+def index(request: HttpRequest, launch_id: str):
+    try:
+        message_launch = get_message_launch(request, launch_id)
+        launch_data = message_launch.get_launch_data()
+        lis = launch_data['https://purl.imsglobal.org/spec/lti/claim/lis']
+        course_sis_id = lis['course_offering_sourcedid']
+        user_sis_id = lis['person_sourcedid']
+        extra = {
+            'course_sis_id': course_sis_id,
+            'user_sis_id': user_sis_id,
+            'launch_data': launch_data,
+        }
+    except Exception as e:
+        logger.exception(f'Failed to launch self-unenroll tool: {e}', extra=extra)
+        return render(request, 'self_unenrollment_tool/error.html', {'message': 'There was a problem launching the tool.'})
+
+    if request.method == 'GET':
+        return _get_index_view(request, course_sis_id=course_sis_id, user_sis_id=user_sis_id)
+
+    elif request.method == 'POST':
+        return _handle_unenrollment_post(request, course_sis_id=course_sis_id, user_sis_id=user_sis_id, launch_id=launch_id)
+
+
+def _get_index_view(request, course_sis_id, user_sis_id):
+    template_context = {}
+
+    # make sure this is a self-enroll course
+    self_enrollments = []
+    is_self_enroll_course = False
+    is_self_enrolled = False
+
+    if course_sis_id:
+        se_configs = SelfEnrollmentCourse.objects.filter(course_instance_id=course_sis_id)
+        for c in se_configs:
+            logger.debug(f'Self-enrollment is active for role {c.role_id} (link UUID: {c.uuid})')
+            is_self_enroll_course = True
+
+        if is_self_enroll_course:
+            if user_sis_id:
+                self_enrollments = _get_self_enrollments(course_sis_id=course_sis_id, user_sis_id=user_sis_id)
+                if self_enrollments:
+                    is_self_enrolled = True
+
+    template_context['is_self_enroll_course'] = is_self_enroll_course
+    template_context['is_self_enrolled'] = is_self_enrolled
+    template_context['self_enrollments'] = self_enrollments
+    template_context['self_enrollment_count'] = len(self_enrollments)
+
+    return render(request, 'self_unenrollment_tool/index.html', template_context)
+
+
+def _handle_unenrollment_post(request: HttpRequest, course_sis_id, user_sis_id, launch_id):
+    log_extra = {
+        'course_sis_id': course_sis_id,
+        'user_sis_id': user_sis_id,
+        'launch_id': launch_id,
+    }
+    # first, get all of the user's Canvas enrollments
+    api_response = list_enrollments_courses(request_ctx=SDK_CONTEXT, course_id=f'sis_course_id:{course_sis_id}', user_id=f'sis_user_id:{user_sis_id}')
+    if api_response.status_code == 200:
+        canvas_enrollments = api_response.json()
+        for ce in canvas_enrollments:
+            logger.debug(ce)
+    else:
+        logger.error(f'bad response from Canvas API: {api_response.text}', extra=log_extra)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It looks like you are trying to use `QuerySet.extra` for constructing raw SQL queries using external inputs. This could make your code vulnerable to SQL injection attacks, which can lead to loss of critical data. We recommend you to use `params` or `select_params` for passing parameters to the query string.

[Learn more](https://cwe.mitre.org/data/definitions/89.html)

Similar issue at line numbers 129, 133, 136, 145, 148, and 151.